src/include/transaction/deferred_action_manager.h

```diff
@@ -62,17 +63,21 @@ class DeferredActionManager {
    * Clear the queue and apply as many actions as possible
    * @return numbers of deferred actions processed
    */
-  uint32_t Process(transaction::timestamp_t oldest_txn) {
+  uint32_t Process() {
+    timestamp_manager_->CheckOutTimestamp();
+    const transaction::timestamp_t oldest_txn = timestamp_manager_->OldestTransactionStartTime();
     // Check out a timestamp from the transaction manager to determine the progress of
     // running transactions in the system.
     const auto backlog_size = static_cast<uint32_t>(back_log_.size());
     uint32_t processed = ClearBacklog(oldest_txn);
     // There is no point in draining new actions if we haven't cleared the backlog.
     // This leaves some mechanisms for the rest of the system to detect congestion
     // at the deferred action manager and potentially backoff
-    if (backlog_size != processed) return processed;
-    // Otherwise, ingest all the new actions
-    processed += ProcessNewActions(oldest_txn);
+    if (backlog_size == processed) {
+      // ingest all the new actions
+      processed += ProcessNewActions(oldest_txn);
```

Comment:
We should leave a TODO that this is now more conservative than it needs to be and can artificially delay garbage collection.  We should be able to query the cached oldest transaction (should be cheap) in between each event and more aggressively clear the backlog.

Reply:
I can try benchmarking that this weekend also.